lumberjack/src/span.rs

```diff
@@ -4,76 +4,89 @@ use std::ops::Range;
+    pub end: usize,
+    skips: Option<HashSet<usize>>,
 }
 
 impl From<usize> for Span {
     fn from(idx: usize) -> Self {
-        Span::new_continuous(idx, idx + 1)
-    }
-}
-
-impl Clone for Span {
-    fn clone(&self) -> Self {
-        match self {
-            Span::Continuous(span) => Span::Continuous(*span),
-            Span::Discontinuous(span) => Span::Discontinuous(span.clone()),
+        Span {
+            start: idx,
+            end: idx + 1,
+            skips: None,
         }
     }
 }
 
 impl Span {
-    /// Return whether `index` is inside the `Span`.
-    pub fn contains(&self, index: usize) -> bool {
-        match self {
-            Span::Continuous(span) => span.contains(index),
-            Span::Discontinuous(span) => span.contains(index),
+    /// Create new span with skipped indices.
+    ///
+    /// `skips` has to be non-empty, to construct a span without skips, use `Span::new`.
+    pub(crate) fn new_with_skips(lower: usize, upper: usize, skips: HashSet<usize>) -> Self {
+        assert!(lower < upper, "Span start has to be smaller then end.");
+        assert!(
+            !skips.is_empty(),
+            "Skips have to be non-empty if using this constructor."
+        );
+        assert_ne!(upper - lower - skips.len(), 0, "Can't skip all indices.");
+        Span {
+            start: lower,
+            end: upper,
+            skips: Some(skips),
         }
     }
 
-    /// Get this spans lower bounds.
-    pub fn lower(&self) -> usize {
-        match self {
-            Span::Continuous(span) => span.lower,
-            Span::Discontinuous(span) => span.lower,
+    /// Create new continuous span.
+    pub(crate) fn new(lower: usize, upper: usize) -> Self {
+        assert!(lower < upper, "Span start has to be smaller then end.");
+        Span {
+            start: lower,
+            end: upper,
+            skips: None,
         }
     }
 
-    /// Get this spans upper bounds.
-    pub fn upper(&self) -> usize {
-        match self {
-            Span::Continuous(span) => span.upper,
-            Span::Discontinuous(span) => span.upper,
+    /// Return whether then span covers the index.
+    pub fn contains(&self, index: usize) -> bool {
+        if self.start <= index && self.end > index {
+            self.skips
+                .as_ref()
+                .map(|skips| !skips.contains(&index))
+                .unwrap_or(true)
+        } else {
+            false
         }
     }
 
     /// Get this spans bounds as a tuple.
     pub fn bounds(&self) -> (usize, usize) {
-        (self.lower(), self.upper())
+        (self.start, self.end)
     }
 
     /// Get the number of indices covered.
     pub fn n_indices(&self) -> usize {
-        match self {
-            Span::Discontinuous(span) => span.upper - span.lower - span.skip.len(),
-            Span::Continuous(span) => span.upper - span.lower,
+        if let Some(ref skips) = self.skips {
+            self.end - self.start - skips.len()
```

Comment:
Is this subject to changes, or could `n_indices` be set at construction time?

Reply:
Spans can/will change